sympy/core/symbol.py

```diff
@@ -600,6 +600,53 @@ def matches(self, expr, repl_dict=None, old=False):
         return repl_dict
 
 
+class CodeblockResult(Symbol):
```

Comment:
This should go in the codegen module.

Reply:
Another option is to make this a node that stores both a result expression and a codeblock. Then doblocks could just return this one thing, and you would access them with

```py
result_expression, block = codeblock_result.args # result_expression is a Dummy, block is a CodeBlock
```

Otherwise, I don't really see why a separate class is needed here at all. We can just return the Dummy that is used in the printer.

Reply:
I agree, this should be in the codegen module. And although we could get away with a plain dummy, I like the idea of a new `CodeblockResult` for several reasons. First, I like being able to recover all the replaced subexpressions by looking for the `CodeblockResult` objects in an expression. Also, I intent to keep the original subexpression and codeblock stored in the `CodeblockResult` object to keep `new_expr` completely self-contained and be able to recover the full original expression from the new expression. 

I can see something like this being used to debug code generation but it might not be necessary.

Reply:
Maybe what would make more sense would be for the codeblock objects themselves to store the results. It could default to the last expression in the block, but could be overwritten by something else. 

I think this sort of thing is necessary to keep track of which result expression corresponds to which code block. If you generate code for many sums, for example, which depend on each other, you need to keep them in order or the code won't work.

> I like the idea of being able to manipulate each code block individually. Specially if we are thinking to do any sort of post processing in doblocks(). For example, you might want to do CSE jointly only on a subset of code blocks using a keyword argument.

You also need to make sure that the blocks stay in order here, even after doing cse. 

> This aims to recover middle steps from the blocks? I like the feature but I would add it as a third optional variable to unpack when returning the output.

No the idea is just that you could have something that returns multiple values (in Python you would just use a `tuple`, but that sort of thing doesn't exist in other languages).